libbeat/tests/compose/compose.go

```diff
@@ -0,0 +1,249 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Wait for health
+	err = compose.Wait(timeout, services...)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// Start the container, unless it's running already
+func (c *composeProject) Start(service string) error {
+	servicesStatus, err := c.getServices(service)
+	if err != nil {
+		return err
+	}
+
+	if servicesStatus[service] != nil {
+		if servicesStatus[service].Running {
+			// Someone is running it
+			return nil
+		}
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	return c.p.Up(context.Background(), options.Up{
+		Create: options.Create{
+			ForceBuild: true,
+		},
+	}, service)
+}
+
+// Ensure all wanted services are healthy. Wait loop (60s timeout)
+func (c *composeProject) Wait(seconds int, services ...string) error {
+	healthy := false
+	for !healthy && seconds > 0 {
+		healthy = true
+
+		servicesStatus, err := c.getServices(services...)
+		if err != nil {
+			return err
+		}
+
+		for _, s := range servicesStatus {
+			if !s.Healthy {
+				healthy = false
+				break
+			}
+		}
+
+		time.Sleep(1 * time.Second)
+		seconds--
+	}
+
+	if !healthy {
+		return errors.New("Timeout waiting for services to be healthy")
+	}
+	return nil
+}
+
+func (c *composeProject) Kill(service string) {
+	c.Lock()
+	defer c.Unlock()
+
+	c.p.Kill(context.Background(), "KILL", service)
+}
+
+func (c *composeProject) KillOld(except []string) error {
+	// Do not kill ourselves or elasticsearch :)
+	except = append(except, "beat", "elasticsearch")
+
+	servicesStatus, err := c.getServices()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range servicesStatus {
+		// Ignore the ones we want
+		if contains(except, s.Name) {
+			continue
+		}
+
+		if s.Old {
+			c.Kill(s.Name)
+		}
+	}
+
+	return nil
+}
+
+// Lock acquires the lock (30s) timeout
+func (c *composeProject) Lock() {
+	seconds := 30
+	for seconds > 0 {
+		file, err := os.OpenFile(c.file+".lock", os.O_CREATE|os.O_EXCL, 0500)
+		file.Close()
+		if err != nil {
+			fmt.Println("docker-compose.yml is locked, waiting")
+			time.Sleep(1 * time.Second)
+			seconds--
+			continue
+		}
+		return
+	}
+
+	// This should rarely happen as we lock for start only, less than a second
+	panic(errors.New("Timeout waiting for lock, please remove docker-comose.yml.lock"))
+}
+
+func (c *composeProject) Unlock() {
+	os.Remove(c.file + ".lock")
+}
+
+func (c *composeProject) getServices(filter ...string) (map[string]*serviceInfo, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	result := make(map[string]*serviceInfo)
+	services, err := c.p.Ps(context.Background(), filter...)
+	if err != nil {
+		return nil, err
+	}
+
+	containers, err := c.p.Containers(context.Background(), project.Filter{State: project.Running}, filter...)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range services {
+		name := strings.Split(c["Name"], "_")[1]
+		// In case of several (stopped) containers, always prefer info about running ones
+		if result[name] != nil {
+			if result[name].Running {
```

Comment:
You comment above that you are looking for the running containers but here you seem to skip the running once?

Reply:
I'm checking in the result array, if it already has a Running container we ignore the current one (`c`)